packages/flutter_tools/lib/src/vmservice.dart

```diff
@@ -40,155 +32,543 @@ class VMService {
   }
 
-  StreamController<Event> _getEventController(String eventName) {
-    StreamController<Event> controller = _eventControllers[eventName];
+  StreamController<ServiceEvent> _getEventController(String eventName) {
+    StreamController<ServiceEvent> controller = _eventControllers[eventName];
     if (controller == null) {
-      controller = new StreamController<Event>.broadcast();
+      controller = new StreamController<ServiceEvent>.broadcast();
       _eventControllers[eventName] = controller;
     }
     return controller;
   }
 
   void _handleStreamNotify(Map<String, dynamic> data) {
-    Event event = new Event(data['event']);
-    _getEventController(data['streamId']).add(event);
+    final String streamId = data['streamId'];
+    final Map<String, dynamic> eventData = data['event'];
+    final Map<String, dynamic> eventIsolate = eventData['isolate'];
+    ServiceEvent event;
+    if (eventIsolate != null) {
+      // getFromMap creates the Isolate if necessary.
+      Isolate isolate = vm.getFromMap(eventIsolate);
+      event = new ServiceObject._fromMap(isolate, eventData);
+      if (event.kind == ServiceEvent.kIsolateExit) {
+        vm._isolateCache.remove(isolate.id);
+        vm._buildIsolateList();
+      } else if (event.kind == ServiceEvent.kIsolateRunnable) {
+        // Force reload once the isolate becomes runnable so that we
+        // update the root library.
+        isolate.reload();
+      }
+    } else {
+      event = new ServiceObject._fromMap(vm, eventData);
+    }
+    _getEventController(streamId).add(event);
   }
 
-  Future<Null> populateIsolateInfo() async {
-    // Calling this has the side effect of populating the isolate information.
-    await waitFirstIsolate;
+  Future<Null> _streamListen(String streamId) async {
+    if (!_listeningFor.contains(streamId)) {
+      _listeningFor.add(streamId);
+      peer.sendRequest('streamListen',
+                       <String, dynamic>{ 'streamId': streamId });
+    }
   }
 
-  Future<IsolateRef> get waitFirstIsolate async {
-    if (isolates.isNotEmpty)
-      return isolates.first;
+  /// Returns the VM and always reloads.
+  Future<VM> getVM() {
+    return _vm.reload();
+  }
+}
 
-    _waitFirstIsolateCompleter ??= new Completer<IsolateRef>();
+/// An error that is thrown when constructing/updating a service object.
+class VMServiceObjectLoadError {
+  VMServiceObjectLoadError(this.message, this.map);
+  final String message;
+  final Map<String, dynamic> map;
+}
 
-    getVM().then((VM vm) {
-      for (IsolateRef isolate in vm.isolates)
-        _addIsolate(isolate);
-    });
+bool _isServiceMap(Map<String, dynamic> m) {
+  return (m != null) && (m['type'] != null);
+}
+bool _hasRef(String type) => (type != null) && type.startsWith('@');
+String _stripRef(String type) => (_hasRef(type) ? type.substring(1) : type);
 
-    return _waitFirstIsolateCompleter.future;
+void _upgradeCollection(dynamic collection,
```

Comment:
This is the magic function which recursively walks the raw response from the service protocol and upgrades values in the map into full service objects. It's an implementation detail but I'll add some documentation here.